go/src/meguca/geoip/db.go

```diff
@@ -4,38 +4,90 @@ import (
 	"net"
 	"strings"
 	"time"
+	"os"
+	"errors"
+	"fmt"
+	"sync"
+	"io/ioutil"
+	"net/http"
+	"crypto/md5"
+	"path/filepath"
+
+	"meguca/db"
 
 	"github.com/oschwald/maxminddb-golang"
 	"github.com/go-playground/log"
+	"github.com/mholt/archiver"
 )
 
 // nil, if database not loaded
-var db *maxminddb.Reader
+var (
+	// Ensure we only load the old GeoIP DB once on server start
+	once sync.Once
+	// Ensures no data races
+	rw sync.RWMutex
+	// GeoIP database
+	gdb *maxminddb.Reader
+)
 
+// NY location
 var NY *time.Location
 
 func init() {
 	NY, _ = time.LoadLocation("America/New_York")
 }
 
+// Load checks if the GeoLite DB exists, and calls load it if it does
 func Load() error {
-	var err error
-	db, err = maxminddb.Open("GeoLite2-Country.mmdb")
+	go func() {
+		if err := check(); err != nil {
+			rw.Lock()
+			defer rw.Unlock()
+			gdb = nil
+			log.Warn("Unable to use GeoLite DB: ", err)
+		}
+	}()
+
+	return nil
+}
+
+// loads the GeoLite DB
+func load() (err error) {
+	rw.Lock()
+	defer rw.Unlock()
+	gdb, err = maxminddb.Open("GeoLite2-Country.mmdb")
+	return
+}
+
+// reloads the GeoLite DB
+func reload() error {
+	rw.Lock()
+	defer rw.Unlock()
+
+	if gdb == nil {
+		return load()
+	}
+
+	err := gdb.Close()
+
 	if err != nil {
-		db = nil
+		return err
 	}
-	return nil
+
+	gdb = nil
+	return load()
 }
 
-// Look up the country ISO code of the IP
+// LookUp looks up the country ISO code of the IP
 func LookUp(ip string) (iso string) {
 	// DB not loaded
-	if db == nil {
+	if gdb == nil {
```

Comment:
Nonono, you need to keep the RLock() here. The problem I had is you using both and Rlock() and Lock() in this same function for no reason. The RLock() protects the gdb pointer itself, not the data structure it points to.